packages/data-context/src/actions/ProjectActions.ts

```diff
@@ -462,4 +473,166 @@ export class ProjectActions {
       await this.ctx.actions.wizard.scaffoldTestingType()
     }
   }
+
+  async runSpec ({ specPath }: { specPath: string}) {
+    const waitForBrowserToOpen = async () => {
+      const browserStatusSubscription = this.ctx.emitter.subscribeTo('browserStatusChange', { sendInitial: false })
+
+      // Wait for browser to finish launching. Browser is either launched from scratch
+      // or relaunched when switching testing types - we need to wait in either case
+      // We wait a maximum of 3 seconds so we don't block indefinitely in case something
+      // goes sideways with the browser launch process. This is broken up into three
+      // separate 'waits' in case we have to watch a browser relaunch (close > opening > open)
+      debug('Waiting for browser to report `open`')
+      let maxIterations = 3
+
+      while (this.ctx.coreData.app.browserStatus !== 'open') {
+        await Promise.race([
+          new Promise((resolve) => setTimeout(resolve, 1000)),
+          browserStatusSubscription.next(),
+        ])
+
+        if (--maxIterations === 0) {
+          break
+        }
+      }
+
+      browserStatusSubscription.return(undefined as any)
+    }
+
+    try {
+      if (!this.ctx.currentProject) {
+        throw new RunSpecError('NO_PROJECT', 'A project must be open prior to attempting to run a spec')
+      }
+
+      if (!specPath) {
+        throw new RunSpecError('NO_SPEC_PATH', '`specPath` must be a non-empty string')
+      }
+
+      let targetTestingType: TestingType
+
+      // Check to see whether input specPath matches the specPattern for one or the other testing type
+      // If it matches neither then we can't run the spec and we should error
+      if (await this.ctx.project.matchesSpecPattern(specPath, 'e2e')) {
```

Comment:
Just to note here - it's possible that the spec path could match the spec patterns for both testing types. In this case, we assume that it's e2e. This appears to be how the app behaves today when choosing which specs to display in the specs list.

Reply:
Good callout - this should match existing logic which basically stipulates that e2e captures any specs that match its pattern and ct captures any specs that match its pattern and don't match e2e's